microservice/src/test/java/de/gsi/microservice/concepts/cmwlight/CmwLightClient.java

```diff
@@ -0,0 +1,532 @@
+    public static final String FILTERS_TAG = "c";
+    public static final String DATA_TAG = "x";
+    public static final String SESSION_ID_TAG = "d";
+    public static final String SESSION_BODY_TAG = "e";
+    public static final String PROPERTY_NAME_TAG = "f";
+
+    // request type used in request header REQ_TYPE_TAG
+    public static final byte RT_GET = 0;
+    public static final byte RT_SET = 1;
+    public static final byte RT_CONNECT = 2;
+    public static final byte RT_REPLY = 3;
+    public static final byte RT_EXCEPTION = 4;
+    public static final byte RT_SUBSCRIBE = 5;
+    public static final byte RT_UNSUBSCRIBE = 6;
+    public static final byte RT_NOTIFICATION_DATA = 7;
+    public static final byte RT_NOTIFICATION_EXC = 8;
+    public static final byte RT_SUBSCRIBE_EXCEPTION = 9;
+    public static final byte RT_EVENT = 10; // Also used as close
+    public static final byte RT_SESSION_CONFIRM = 11;
+
+    // UpdateType
+    public static final byte UT_NORMAL = (byte) 0;
+    public static final byte UT_FIRST_UPDATE = (byte) 1; // Initial update sent when the subscription is created.
+    public static final byte UT_IMMEDIATE_UPDATE = (byte) 2; // Update sent after the value has been modified by a set call.
+
+    public enum ConnectionState {
+        DISCONNECTED,
+        CONNECTING,
+        CONNECTED
+    }
+
+    public CmwLightClient(String address) {
+        controlChannel = context.createSocket(SocketType.DEALER);
+        controlChannel.setIdentity(getIdentity().getBytes()); // hostname/process/id/channel
+        controlChannel.setSndHWM(0);
+        controlChannel.setRcvHWM(0);
+        controlChannel.setLinger(0);
+        this.address = address;
+    }
+
+    public ZMQ.Socket getSocket() {
+        return controlChannel;
+    }
+
+    private String getIdentity() {
+        String hostname;
+        try {
+            hostname = InetAddress.getLocalHost().getHostName();
+        } catch (UnknownHostException e) {
+            hostname = "localhost";
+        }
+        final long processId = ProcessHandle.current().pid();
+        final int connectionId = CmwLightClient.connectionId.incrementAndGet();
+        final int channelId = this.channelId.incrementAndGet();
+        return hostname + '/' + processId + '/' + connectionId + '/' + channelId;
+    }
+
+    public void connect() throws RdaLightException {
+        controlChannel.connect(address);
+        if (connectionState.getAndSet(ConnectionState.CONNECTING) != ConnectionState.DISCONNECTED) {
+            return;
+        }
+        controlChannel.send(new byte[] { CLIENT_CONNECT }, ZMQ.SNDMORE);
+        controlChannel.send("1.0.0".getBytes());
+        final ZMsg conAck = ZMsg.recvMsg(controlChannel);
+        if (!Arrays.equals(conAck.pollFirst().getData(), new byte[] { SERVER_CONNECT_ACK })) {
+            throw new RdaLightException(conAck.toString());
+        }
+    }
+
+    public Reply receiveData() throws RdaLightException {
```

Comment:
Codacy found an issue: [The method 'receiveData()' has an NPath complexity of 28800, current threshold is 200](https://app.codacy.com/gh/GSI-CS-CO/chart-fx/pullRequest?prid=6417799)